src/main/kotlin/net/atos/zac/identity/IdentityService.kt

```
@@ -4,197 +4,58 @@
-            )
-        }
-    }
-
-    private fun readAttributeToString(attributes: Attributes, attributeName: String): String? {
-        val attribute = attributes[attributeName]
-        return attribute?.get()?.toString()
-    }
-
-    private fun readAttributeToListOfStrings(attributes: Attributes): List<String> =
-        attributes[GROUP_MEMBER_ATTRIBUTE]?.all?.toList()?.map {
-            it.toString()
-        } ?: emptyList()
-
-    private fun Attributes.toUser(): User {
-        val userID = readAttributeToString(this, USER_ID_ATTRIBUTE)
-        require(userID != null) { "User ID is required" }
-        return User(
-            id = userID,
-            firstName = readAttributeToString(this, USER_FIRST_NAME_ATTRIBUTE),
-            lastName = readAttributeToString(this, USER_LAST_NAME_ATTRIBUTE),
-            email = readAttributeToString(this, USER_MAIL_ATTRIBUTE)
-        )
-    }
-
-    private fun Attributes.toGroup(): Group {
-        val groupID = readAttributeToString(this, GROUP_ID_ATTRIBUTE)
-        require(groupID != null) { "Group ID is required" }
-        return Group(
-            id = groupID,
-            name = readAttributeToString(this, GROUP_NAME_ATTRIBUTE) ?: groupID,
-            email = readAttributeToString(this, GROUP_MAIL_ATTRIBUTE)
-        )
+    fun listUsers(): List<User> = keycloakZacRealmResource.users()
+        .list()
+        .map { it.toUser() }
+        .sortedBy { it.fullName }
+
+    fun listGroups(): List<Group> = keycloakZacRealmResource.groups()
+        // retrieve groups with 'full representation' or else the group attributes will not be filled
+        .groups("", 0, Integer.MAX_VALUE, false)
+        .map { it.toGroup() }
+        .sortedBy { it.name }
+
+    fun readUser(userId: String): User = keycloakZacRealmResource.users()
+        .search(userId).map { it.toUser() }.firstOrNull()
+        // is this fallback really needed? better to return null or throw a custom exception
+        ?: User(userId)
```

i would also throw an error or return null here as the user is not found

---

similar to the other places where you do the same

---

yep. that's why I have added this todo. I suggest to look into this when we implement this for real. This fallback (which is also present in the current code) is really weird if you ask me.